llvm/include/llvm/Analysis/DomConditionCache.h

```diff
@@ -18,18 +18,34 @@
 #define LLVM_ANALYSIS_DOMCONDITIONCACHE_H
 
 #include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/BitmaskEnum.h"
 #include "llvm/ADT/DenseMap.h"
 #include "llvm/ADT/SmallVector.h"
+#include <cstdint>
 
 namespace llvm {
 
 class Value;
 class BranchInst;
 
+enum class DomConditionFlag : uint8_t {
+  None = 0,
+  KnownBits = 1 << 0,
+  KnownFPClass = 1 << 1,
+  PowerOfTwo = 1 << 2,
+  ICmp = 1 << 3,
+};
+
+LLVM_DECLARE_ENUM_AS_BITMASK(
+    DomConditionFlag,
+    /*LargestValue=*/static_cast<uint8_t>(DomConditionFlag::ICmp));
+
 class DomConditionCache {
 private:
   /// A map of values about which a branch might be providing information.
-  using AffectedValuesMap = DenseMap<Value *, SmallVector<BranchInst *, 1>>;
+  using AffectedValuesMap =
```

Comment:
as it seems like conditionsFor is called for one DomConditionFlag value maybe better add the flag as en argument and change map to store the flag as part of the key to the map?
DenseMap<std::pair<Value *, DomConditionFlag>, SmallVector<BranchInst *, 1>>;

Reply:
Compile-time impact looks worse: http://llvm-compile-time-tracker.com/compare.php?from=6328e0ff42d29db0923615afd8430b4518e8ecbe&to=4b0feb47aec9e3a854610e8a223511e7f830f0b1&stat=instructions:u

Reply:
hmm noticed now that you mentioned that it was the calls to registerBranch that was the problem in  https://github.com/llvm/llvm-project/pull/117442#discussion_r1855839736 and this will result in more work then.
but if it is the registerBranch that is the problem I assume that any filtering added will result in worse compile time.